text/0080-simulation-task-groups.md

```diff
@@ -0,0 +1,128 @@
+For instance, when testing a stream component, the output stream must typically be read in parallel with writing the input stream to avoid backpressure from the output stream to propagate back to the input stream and deadlock the simulation.
+Currently this must be done by adding independent testbenches, which can be awkward to synchronize.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Typical testbenches for a stream component can currently look like this:
+```python
+test_vectors = [...]
+
+async def input_testbench(ctx: SimulatorContext):
+    for input in test_vectors:
+        await send_packet(ctx, dut.i, input)
+
+async def output_testbench(ctx: SimulatorContext):
+    for input in test_vectors:
+        output = await recv_packet(ctx, dut.o)
+        assert output == expected_output(input)
+```
+
+With task groups, this can instead be written like:
+```python
+test_vectors = [...]
+
+async def testbench(ctx: SimulatorContext):
+    for input in test_vectors:
+        async with ctx.group() as group:
+            group.start(send_packet(ctx, dut.i, input))
+            output = await recv_packet(ctx, dut.o)
+            assert output == expected_output(input)
+```
+
+In a similar manner to background testbenches, it is also possible to add background tasks, for tasks that are not intended to run to completion.
+This allows code like this:
+```python
+@asynccontextmanager
+async def timeout(ctx, ticks, domain = 'sync'):
+    async def task():
+        await ctx.tick(domain).repeat(ticks) # Never returns if the task group ends before `ticks` have elapsed.
+        raise TimeoutError()
+
+    async with ctx.group() as group:
+        group.start(task(ctx, ticks, domain), background = True)
+        yield
+
+async def testbench(ctx):
+    async with timeout(ctx, 100):
+        ... # Some operation expected to take less than 100 ticks
+
+    async with timeout(ctx, 200):
+        ... # Some other operation expected to take less than 200 ticks
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`SimulatorContext` have the following methods added:
+- `group() -> TaskGroup`
+  - Create a new task group.
+- `async gather(coros*) -> tuple`
+  - Shorthand for creating a task group, starting all `coros`, letting the group run to completion and collecting the return values.
+  - Example implementation:
+    ```python
+        async def gather(self, *coros):
+            async with self.group() as group:
+                tasks = [group.start(coro) for coro in coros]
+            return tuple(task.result() for task in tasks)
+    ```
+
+`TaskGroup` is added with the following methods:
+- `start(coro, *, background = False) -> Task`
+  - Create and start a new task.
+  - A background task can be temporarily made non-background with `ctx.critical()` like a regular testbench.
```

Comment:
I think the text needs to clearly state that starting tasks is only possible between `__aenter__` and `__aexit__`.